VocaDbWeb/Scripts/Stores/Discussion/DiscussionIndexStore.ts

```diff
@@ -0,0 +1,185 @@
+import DiscussionFolderContract from '@DataContracts/Discussion/DiscussionFolderContract';
+import DiscussionTopicContract from '@DataContracts/Discussion/DiscussionTopicContract';
+import LoginManager from '@Models/LoginManager';
+import DiscussionRepository from '@Repositories/DiscussionRepository';
+import ServerSidePagingStore from '@Stores/ServerSidePagingStore';
+import _ from 'lodash';
+import {
+	action,
+	makeObservable,
+	observable,
+	reaction,
+	runInAction,
+} from 'mobx';
+
+import DiscussionTopicEditStore from './DiscussionTopicEditStore';
+import DiscussionTopicStore from './DiscussionTopicStore';
+
+export default class DiscussionIndexStore {
+	@observable public folders: DiscussionFolderContract[] = [];
+	@action public setFolders = (value: DiscussionFolderContract[]): void => {
+		this.folders = value;
+	};
+
+	@observable public newTopic: DiscussionTopicEditStore;
+	@action public setNewTopic = (value: DiscussionTopicEditStore): void => {
+		this.newTopic = value;
+	};
+
+	public readonly paging = new ServerSidePagingStore(30); // Paging store
+
+	@observable public recentTopics: DiscussionTopicContract[] = [];
+	@action public setRecentTopics = (value: DiscussionTopicContract[]): void => {
+		this.recentTopics = value;
+	};
+
+	@observable public selectedFolder?: DiscussionFolderContract = undefined;
+	@action public setSelectedFolder = (
+		value?: DiscussionFolderContract,
+	): void => {
+		this.selectedFolder = value;
+	};
+
+	@observable public selectedTopic?: DiscussionTopicStore = undefined;
+	@action public setSelectedTopic = (value?: DiscussionTopicStore): void => {
+		this.selectedTopic = value;
+	};
+
+	@observable public showCreateNewTopic: boolean = false;
+	@action public setShowCreateNewTopic = (value: boolean): void => {
+		this.showCreateNewTopic = value;
+	};
+
+	@observable public topics: DiscussionTopicContract[] = [];
+	@action public setTopics = (value: DiscussionTopicContract[]): void => {
+		this.topics = value;
+	};
+
+	public constructor(
+		public readonly loginManager: LoginManager,
+		private readonly discussionRepo: DiscussionRepository,
+		private readonly canDeleteAllComments: boolean,
+	) {
+		makeObservable(this);
+
+		this.newTopic = new DiscussionTopicEditStore(loginManager, this.folders);
+
+		discussionRepo.getFolders({}).then((folders) => {
+			this.setFolders(folders);
+		});
+
+		discussionRepo.getTopics({}).then((result) => {
+			this.setRecentTopics(result.items);
+		});
+
+		reaction(
+			() => this.selectedFolder,
+			(folder) => {
+				this.showCreateNewTopic = false;
+				this.selectedTopic = undefined;
+				this.paging.goToFirstPage();
+
+				this.loadTopics(folder);
+			},
+		);

```

Comment:
The callback of `reaction` is *not* called if the new value is equal to the old value, while the callback of `subscribe` is called. We need to do something if we want the exact same behavior with Knockout's subscribe.

Let's say topic 39 is located in the folder 1.

| Knockout | MobX |
| --- | --- |
| Go to `/discussion/topics/39`. | Go to `/discussion/topics/39`. |
| The `selectedFolder` is now folder 1. | The `selectedFolder` is now folder 1. |
| Click on the folder name on the breadcrumb. | Click on the folder name on the breadcrumb |
| The route has changed, and `selectFolderById` is called. | The route has changed, and `selectFolderById` is called. |
| The `selectedFolder` is set to folder 1 (unchanged). | The `selectedFolder` is set to folder 1 (unchanged). |
| The callback of `subscribe` is called. (old: folder 1, new: folder 1) | The callback of `reaction` is *not* called. |